File: core-lib/bindings-swift/core_lib.swift
import Foundation

#if canImport(core_libFFI)
import core_libFFI
#endif

public struct Client: Equatable, Hashable {
    public var id: String
    public var organizationId: UInt64
    public var name: String
    public var email: String
    public var phone: String
    public var address: String
    public var city: String
    public var country: String
    public var createdAt: String

    public init(id: String, organizationId: UInt64, name: String, email: String, phone: String, address: String, city: String, country: String, createdAt: String) {
        self.id = id
        self.organizationId = organizationId
        self.name = name
        self.email = email
        self.phone = phone
        self.address = address
        self.city = city
        self.country = country
        self.createdAt = createdAt
    }
}

#if compiler(>=6)
extension Client: Sendable {}
#endif

public struct InvoiceItem: Equatable, Hashable {
    public var description: String
    public var quantity: UInt32
    public var price: Double

    public init(description: String, quantity: UInt32, price: Double) {
        self.description = description
        self.quantity = quantity
        self.price = price
    }
}

#if compiler(>=6)
extension InvoiceItem: Sendable {}
#endif

public enum InvoiceStatus: Equatable, Hashable {
    case draft
    case pending
    case paid
    case overdue

    init(core: CoreInvoiceStatus) {
        switch core {
        case CoreInvoiceStatusDraft:
            self = .draft
        case CoreInvoiceStatusPending:
            self = .pending
        case CoreInvoiceStatusPaid:
            self = .paid
        case CoreInvoiceStatusOverdue:
            self = .overdue
        default:
            self = .draft
        }
    }
}

#if compiler(>=6)
extension InvoiceStatus: Sendable {}
#endif

public struct Invoice: Equatable, Hashable {
    public var id: String
    public var invoiceNumber: String
    public var clientId: String
    public var items: [InvoiceItem]
    public var taxRate: Double
    public var discount: Double
    public var status: InvoiceStatus
    public var signature: String?
    public var issueDate: String
    public var dueDate: String
    public var currency: String
    public var notes: String
    public var createdAt: String

    public init(id: String, invoiceNumber: String, clientId: String, items: [InvoiceItem], taxRate: Double, discount: Double, status: InvoiceStatus, signature: String?, issueDate: String, dueDate: String, currency: String, notes: String, createdAt: String) {
        self.id = id
        self.invoiceNumber = invoiceNumber
        self.clientId = clientId
        self.items = items
        self.taxRate = taxRate
        self.discount = discount
        self.status = status
        self.signature = signature
        self.issueDate = issueDate
        self.dueDate = dueDate
        self.currency = currency
        self.notes = notes
        self.createdAt = createdAt
    }
}

#if compiler(>=6)
extension Invoice: Sendable {}
#endif

public struct DashboardStats: Equatable, Hashable {
    public var totalRevenue: Double
    public var paidCount: UInt64
    public var pendingCount: UInt64
    public var overdueCount: UInt64
    public var currency: String

    public init(totalRevenue: Double, paidCount: UInt64, pendingCount: UInt64, overdueCount: UInt64, currency: String) {
        self.totalRevenue = totalRevenue
        self.paidCount = paidCount
        self.pendingCount = pendingCount
        self.overdueCount = overdueCount
        self.currency = currency
    }
}

#if compiler(>=6)
extension DashboardStats: Sendable {}
#endif

public struct InvoiceStatusCount: Equatable, Hashable {
    public var id: String
    public var status: String
    public var count: UInt64

    public init(id: String, status: String, count: UInt64) {
        self.id = id
        self.status = status
        self.count = count
    }
}

#if compiler(>=6)
extension InvoiceStatusCount: Sendable {}
#endif

public struct MonthlyRevenue: Equatable, Hashable {
    public var month: String
    public var year: UInt16
    public var currency: String
    public var amount: UInt32

    public init(month: String, year: UInt16, currency: String, amount: UInt32) {
        self.month = month
        self.year = year
        self.currency = currency
        self.amount = amount
    }
}

#if compiler(>=6)
extension MonthlyRevenue: Sendable {}
#endif

public struct Dashboard: Equatable, Hashable {
    public var dashboardStats: DashboardStats
    public var invoicesStatus: [InvoiceStatusCount]
    public var monthlyRevenue: [MonthlyRevenue]

    public init(dashboardStats: DashboardStats, invoicesStatus: [InvoiceStatusCount], monthlyRevenue: [MonthlyRevenue]) {
        self.dashboardStats = dashboardStats
        self.invoicesStatus = invoicesStatus
        self.monthlyRevenue = monthlyRevenue
    }
}

#if compiler(>=6)
extension Dashboard: Sendable {}
#endif

public struct ReferralData: Equatable, Hashable {
    public var totalReferrals: UInt64
    public var activeReferrals: UInt64
    public var totalEarnings: Double
    public var payout: Double
    public var referralCode: String

    public init(totalReferrals: UInt64, activeReferrals: UInt64, totalEarnings: Double, payout: Double, referralCode: String) {
        self.totalReferrals = totalReferrals
        self.activeReferrals = activeReferrals
        self.totalEarnings = totalEarnings
        self.payout = payout
        self.referralCode = referralCode
    }
}

#if compiler(>=6)
extension ReferralData: Sendable {}
#endif

public func fetchClients() -> [Client] {
    let list = core_lib_fetch_clients()
    defer { core_lib_free_clients(list) }
    guard list.len > 0 else { return [] }
    var result: [Client] = []
    result.reserveCapacity(list.len)
    for i in 0..<list.len {
        let c = list.ptr[i]
        result.append(Client(
            id: String(cString: c.id),
            organizationId: c.organization_id,
            name: String(cString: c.name),
            email: String(cString: c.email),
            phone: String(cString: c.phone),
            address: String(cString: c.address),
            city: String(cString: c.city),
            country: String(cString: c.country),
            createdAt: String(cString: c.created_at)
        ))
    }
    return result
}

public func fetchInvoices() -> [Invoice] {
    let list = core_lib_fetch_invoices()
    defer { core_lib_free_invoices(list) }
    guard list.len > 0 else { return [] }
    var result: [Invoice] = []
    result.reserveCapacity(list.len)
    for i in 0..<list.len {
        let inv = list.ptr[i]
        var items: [InvoiceItem] = []
        let itemList = inv.items
        if itemList.len > 0 {
            items.reserveCapacity(itemList.len)
            for j in 0..<itemList.len {
                let it = itemList.ptr[j]
                items.append(InvoiceItem(
                    description: String(cString: it.description),
                    quantity: it.quantity,
                    price: it.price
                ))
            }
        }
        result.append(Invoice(
            id: String(cString: inv.id),
            invoiceNumber: String(cString: inv.invoice_number),
            clientId: String(cString: inv.client_id),
            items: items,
            taxRate: inv.tax_rate,
            discount: inv.discount,
            status: InvoiceStatus(core: inv.status),
            signature: inv.signature.map { String(cString: $0) },
            issueDate: String(cString: inv.issue_date),
            dueDate: String(cString: inv.due_date),
            currency: String(cString: inv.currency),
            notes: String(cString: inv.notes),
            createdAt: String(cString: inv.created_at)
        ))
    }
    return result
}

public func fetchDashboard() -> Dashboard {
    guard let d = core_lib_fetch_dashboard() else {
        return Dashboard(dashboardStats: DashboardStats(totalRevenue: 0, paidCount: 0, pendingCount: 0, overdueCount: 0, currency: ""), invoicesStatus: [], monthlyRevenue: [])
    }
    defer { core_lib_free_dashboard(d) }
    let stats = d.pointee.dashboard_stats
    var statusCounts: [InvoiceStatusCount] = []
    let statusList = d.pointee.invoices_status
    if statusList.len > 0 {
        statusCounts.reserveCapacity(statusList.len)
        for i in 0..<statusList.len {
            let sc = statusList.ptr[i]
            statusCounts.append(InvoiceStatusCount(
                id: String(cString: sc.id),
                status: String(cString: sc.status),
                count: sc.count
            ))
        }
    }
    var monthlyRevenues: [MonthlyRevenue] = []
    let revenueList = d.pointee.monthly_revenue
    if revenueList.len > 0 {
        monthlyRevenues.reserveCapacity(revenueList.len)
        for i in 0..<revenueList.len {
            let mr = revenueList.ptr[i]
            monthlyRevenues.append(MonthlyRevenue(
                month: String(cString: mr.month),
                year: mr.year,
                currency: String(cString: mr.currency),
                amount: mr.amount
            ))
        }
    }
    return Dashboard(
        dashboardStats: DashboardStats(
            totalRevenue: stats.total_revenue,
            paidCount: stats.paid_count,
            pendingCount: stats.pending_count,
            overdueCount: stats.overdue_count,
            currency: String(cString: stats.currency)
        ),
        invoicesStatus: statusCounts,
        monthlyRevenue: monthlyRevenues
    )
}

public func fetchReferralData() -> ReferralData {
    guard let r = core_lib_fetch_referral_data() else {
        return ReferralData(totalReferrals: 0, activeReferrals: 0, totalEarnings: 0, payout: 0, referralCode: "")
    }
    defer { core_lib_free_referral_data(r) }
    return ReferralData(
        totalReferrals: r.pointee.total_referrals,
        activeReferrals: r.pointee.active_referrals,
        totalEarnings: r.pointee.total_earnings,
        payout: r.pointee.payout,
        referralCode: String(cString: r.pointee.referral_code)
    )
}

File: core-lib/bindings-kotlin/CoreLib.kt
@file:Suppress("MemberVisibilityCanBePrivate", "unused", "RedundantVisibilityModifier")

package com.invoice.core_lib

import com.sun.jna.*

// =============================================================================
// Public API
// =============================================================================

public object CoreLib {
    public fun fetchClients(): List<Client> {
        val list = LIB.core_lib_fetch_clients()
        val result = list.ptr?.readClientArray(list.len.toInt())?.map { it.toClient() } ?: emptyList()
        LIB.core_lib_free_clients(list)
        return result
    }

    public fun fetchInvoices(): List<Invoice> {
        val list = LIB.core_lib_fetch_invoices()
        val result = list.ptr?.readInvoiceArray(list.len.toInt())?.map { it.toInvoice() } ?: emptyList()
        LIB.core_lib_free_invoices(list)
        return result
    }

    public fun fetchDashboard(): Dashboard {
        val ptr = LIB.core_lib_fetch_dashboard()
            ?: throw RuntimeException("core_lib_fetch_dashboard returned null")
        val d = CoreDashboard().apply { useMemory(ptr); read() }
        val result = d.toDashboard()
        LIB.core_lib_free_dashboard(ptr)
        return result
    }

    public fun fetchReferralData(): ReferralData {
        val ptr = LIB.core_lib_fetch_referral_data()
            ?: throw RuntimeException("core_lib_fetch_referral_data returned null")
        val r = CoreReferralData().apply { useMemory(ptr); read() }
        val result = r.toReferralData()
        LIB.core_lib_free_referral_data(ptr)
        return result
    }
}

public enum class InvoiceStatus {
    Draft, Pending, Paid, Overdue
}

public data class Client(
    val id: String,
    val organizationId: ULong,
    val name: String,
    val email: String,
    val phone: String,
    val address: String,
    val city: String,
    val country: String,
    val createdAt: String,
)

public data class InvoiceItem(
    val description: String,
    val quantity: UInt,
    val price: Double,
)

public data class Invoice(
    val id: String,
    val invoiceNumber: String,
    val clientId: String,
    val items: List<InvoiceItem>,
    val taxRate: Double,
    val discount: Double,
    val status: InvoiceStatus,
    val signature: String?,
    val issueDate: String,
    val dueDate: String,
    val currency: String,
    val notes: String,
    val createdAt: String,
)

public data class DashboardStats(
    val totalRevenue: Double,
    val paidCount: ULong,
    val pendingCount: ULong,
    val overdueCount: ULong,
    val currency: String,
)

public data class InvoiceStatusCount(
    val id: String,
    val status: String,
    val count: ULong,
)

public data class MonthlyRevenue(
    val month: String,
    val year: UShort,
    val currency: String,
    val amount: UInt,
)

public data class Dashboard(
    val dashboardStats: DashboardStats,
    val invoicesStatus: List<InvoiceStatusCount>,
    val monthlyRevenue: List<MonthlyRevenue>,
)

public data class ReferralData(
    val totalReferrals: ULong,
    val activeReferrals: ULong,
    val totalEarnings: Double,
    val payout: Double,
    val referralCode: String,
)

// =============================================================================
// FFI Layer
// =============================================================================

private val LIB: CoreLibFFI by lazy {
    try {
        System.loadLibrary("core_lib")
    } catch (_: UnsatisfiedLinkError) {
        // fallback for desktop JVM; Native.load will still attempt dlopen
    }
    Native.load("core_lib", CoreLibFFI::class.java)
}

private interface CoreLibFFI : Library {
    fun core_lib_fetch_clients(): CoreClientList
    fun core_lib_free_clients(list: CoreClientList)
    fun core_lib_fetch_invoices(): CoreInvoiceList
    fun core_lib_free_invoices(list: CoreInvoiceList)
    fun core_lib_fetch_dashboard(): Pointer?
    fun core_lib_free_dashboard(dashboard: Pointer?)
    fun core_lib_fetch_referral_data(): Pointer?
    fun core_lib_free_referral_data(referral: Pointer?)
}

// =============================================================================
// JNA Structures — C type mirrors
// =============================================================================

@Structure.FieldOrder("ptr", "len")
private open class CoreClientList : Structure(), Structure.ByValue {
    @JvmField var ptr: Pointer? = null
    @JvmField var len: Long = 0
}

@Structure.FieldOrder("id", "organization_id", "name", "email", "phone", "address", "city", "country", "created_at")
private open class CoreClient : Structure() {
    @JvmField var id: Pointer? = null
    @JvmField var organization_id: Long = 0
    @JvmField var name: Pointer? = null
    @JvmField var email: Pointer? = null
    @JvmField var phone: Pointer? = null
    @JvmField var address: Pointer? = null
    @JvmField var city: Pointer? = null
    @JvmField var country: Pointer? = null
    @JvmField var created_at: Pointer? = null

    fun toClient(): Client = Client(
        id = id?.getString(0) ?: "",
        organizationId = organization_id.toULong(),
        name = name?.getString(0) ?: "",
        email = email?.getString(0) ?: "",
        phone = phone?.getString(0) ?: "",
        address = address?.getString(0) ?: "",
        city = city?.getString(0) ?: "",
        country = country?.getString(0) ?: "",
        createdAt = created_at?.getString(0) ?: ""
    )
}

@Structure.FieldOrder("ptr", "len")
private open class CoreInvoiceItemList : Structure(), Structure.ByValue {
    @JvmField var ptr: Pointer? = null
    @JvmField var len: Long = 0
}

@Structure.FieldOrder("description", "quantity", "price")
private open class CoreInvoiceItem : Structure() {
    @JvmField var description: Pointer? = null
    @JvmField var quantity: Int = 0
    @JvmField var price: Double = 0.0

    fun toInvoiceItem(): InvoiceItem = InvoiceItem(
        description = description?.getString(0) ?: "",
        quantity = quantity.toUInt(),
        price = price
    )
}

@Structure.FieldOrder("ptr", "len")
private open class CoreInvoiceList : Structure(), Structure.ByValue {
    @JvmField var ptr: Pointer? = null
    @JvmField var len: Long = 0
}

@Structure.FieldOrder("id", "invoice_number", "client_id", "items", "tax_rate", "discount", "status", "signature", "issue_date", "due_date", "currency", "notes", "created_at")
private open class CoreInvoice : Structure() {
    @JvmField var id: Pointer? = null
    @JvmField var invoice_number: Pointer? = null
    @JvmField var client_id: Pointer? = null
    @JvmField var items: CoreInvoiceItemList = CoreInvoiceItemList()
    @JvmField var tax_rate: Double = 0.0
    @JvmField var discount: Double = 0.0
    @JvmField var status: Int = 0
    @JvmField var signature: Pointer? = null
    @JvmField var issue_date: Pointer? = null
    @JvmField var due_date: Pointer? = null
    @JvmField var currency: Pointer? = null
    @JvmField var notes: Pointer? = null
    @JvmField var created_at: Pointer? = null

    fun toInvoice(): Invoice = Invoice(
        id = id?.getString(0) ?: "",
        invoiceNumber = invoice_number?.getString(0) ?: "",
        clientId = client_id?.getString(0) ?: "",
        items = items.ptr?.readInvoiceItemArray(items.len.toInt())?.map { it.toInvoiceItem() } ?: emptyList(),
        taxRate = tax_rate,
        discount = discount,
        status = when (status) {
            0 -> InvoiceStatus.Draft
            1 -> InvoiceStatus.Pending
            2 -> InvoiceStatus.Paid
            3 -> InvoiceStatus.Overdue
            else -> InvoiceStatus.Draft
        },
        signature = signature?.getString(0),
        issueDate = issue_date?.getString(0) ?: "",
        dueDate = due_date?.getString(0) ?: "",
        currency = currency?.getString(0) ?: "",
        notes = notes?.getString(0) ?: "",
        createdAt = created_at?.getString(0) ?: ""
    )
}

@Structure.FieldOrder("total_revenue", "paid_count", "pending_count", "overdue_count", "currency")
private open class CoreDashboardStats : Structure(), Structure.ByValue {
    @JvmField var total_revenue: Double = 0.0
    @JvmField var paid_count: Long = 0
    @JvmField var pending_count: Long = 0
    @JvmField var overdue_count: Long = 0
    @JvmField var currency: Pointer? = null

    fun toDashboardStats(): DashboardStats = DashboardStats(
        totalRevenue = total_revenue,
        paidCount = paid_count.toULong(),
        pendingCount = pending_count.toULong(),
        overdueCount = overdue_count.toULong(),
        currency = currency?.getString(0) ?: ""
    )
}

@Structure.FieldOrder("ptr", "len")
private open class CoreInvoiceStatusCountList : Structure(), Structure.ByValue {
    @JvmField var ptr: Pointer? = null
    @JvmField var len: Long = 0
}

@Structure.FieldOrder("id", "status", "count")
private open class CoreInvoiceStatusCount : Structure() {
    @JvmField var id: Pointer? = null
    @JvmField var status: Pointer? = null
    @JvmField var count: Long = 0

    fun toInvoiceStatusCount(): InvoiceStatusCount = InvoiceStatusCount(
        id = id?.getString(0) ?: "",
        status = status?.getString(0) ?: "",
        count = count.toULong()
    )
}

@Structure.FieldOrder("ptr", "len")
private open class CoreMonthlyRevenueList : Structure(), Structure.ByValue {
    @JvmField var ptr: Pointer? = null
    @JvmField var len: Long = 0
}

@Structure.FieldOrder("month", "year", "currency", "amount")
private open class CoreMonthlyRevenue : Structure() {
    @JvmField var month: Pointer? = null
    @JvmField var year: Short = 0
    @JvmField var currency: Pointer? = null
    @JvmField var amount: Int = 0

    fun toMonthlyRevenue(): MonthlyRevenue = MonthlyRevenue(
        month = month?.getString(0) ?: "",
        year = year.toUShort(),
        currency = currency?.getString(0) ?: "",
        amount = amount.toUInt()
    )
}

@Structure.FieldOrder("dashboard_stats", "invoices_status", "monthly_revenue")
private open class CoreDashboard : Structure() {
    @JvmField var dashboard_stats: CoreDashboardStats = CoreDashboardStats()
    @JvmField var invoices_status: CoreInvoiceStatusCountList = CoreInvoiceStatusCountList()
    @JvmField var monthly_revenue: CoreMonthlyRevenueList = CoreMonthlyRevenueList()

    fun toDashboard(): Dashboard = Dashboard(
        dashboardStats = dashboard_stats.toDashboardStats(),
        invoicesStatus = invoices_status.ptr?.readInvoiceStatusCountArray(invoices_status.len.toInt())?.map { it.toInvoiceStatusCount() } ?: emptyList(),
        monthlyRevenue = monthly_revenue.ptr?.readMonthlyRevenueArray(monthly_revenue.len.toInt())?.map { it.toMonthlyRevenue() } ?: emptyList()
    )
}

@Structure.FieldOrder("total_referrals", "active_referrals", "total_earnings", "payout", "referral_code")
private open class CoreReferralData : Structure() {
    @JvmField var total_referrals: Long = 0
    @JvmField var active_referrals: Long = 0
    @JvmField var total_earnings: Double = 0.0
    @JvmField var payout: Double = 0.0
    @JvmField var referral_code: Pointer? = null

    fun toReferralData(): ReferralData = ReferralData(
        totalReferrals = total_referrals.toULong(),
        activeReferrals = active_referrals.toULong(),
        totalEarnings = total_earnings,
        payout = payout,
        referralCode = referral_code?.getString(0) ?: ""
    )
}

// =============================================================================
// Array helpers
// =============================================================================

private fun Pointer.readClientArray(len: Int): Array<CoreClient> {
    val size = CoreClient().size()
    return Array(len) { i ->
        CoreClient().apply {
            useMemory(this@readClientArray.share((i * size).toLong()))
            read()
        }
    }
}

private fun Pointer.readInvoiceItemArray(len: Int): Array<CoreInvoiceItem> {
    val size = CoreInvoiceItem().size()
    return Array(len) { i ->
        CoreInvoiceItem().apply {
            useMemory(this@readInvoiceItemArray.share((i * size).toLong()))
            read()
        }
    }
}

private fun Pointer.readInvoiceArray(len: Int): Array<CoreInvoice> {
    val size = CoreInvoice().size()
    return Array(len) { i ->
        CoreInvoice().apply {
            useMemory(this@readInvoiceArray.share((i * size).toLong()))
            read()
        }
    }
}

private fun Pointer.readInvoiceStatusCountArray(len: Int): Array<CoreInvoiceStatusCount> {
    val size = CoreInvoiceStatusCount().size()
    return Array(len) { i ->
        CoreInvoiceStatusCount().apply {
            useMemory(this@readInvoiceStatusCountArray.share((i * size).toLong()))
            read()
        }
    }
}

private fun Pointer.readMonthlyRevenueArray(len: Int): Array<CoreMonthlyRevenue> {
    val size = CoreMonthlyRevenue().size()
    return Array(len) { i ->
        CoreMonthlyRevenue().apply {
            useMemory(this@readMonthlyRevenueArray.share((i * size).toLong()))
            read()
        }
    }
}
